[jit] fix side-effects and aliasing for custom ops

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#18711 [jit] fix side-effects and aliasing for custom ops**

Previously we didn't track aliasing, mutation, or side effects for
custom ops. This PR adds in guards with the most conservative
assumptions possible: the op will 
1) have side effects, 
2) write to everything
3) produce a wildcard. 

In order to tell whether a given operator is a custom op, this PR introduces 
the concept of a "reserved" namespace (basically all our builtin namespaces).
Custom ops live in non-reserved namespaces, so a check on the namespace 
is sufficient to tell whether a schema/node is "custom" or not.

This is just to get things correct for now. Follow-ups to this:
- Users should be able to specify aliasing/mutability without having to learn
the whole alias annotation schema.
- Relax assumptions a bit. In particular outputs can only alias input tensors, 
they don't have to be wildcards.

Fixes #18490

Differential Revision: [D14730978](https://our.internmc.facebook.com/intern/diff/D14730978)

suo committed:
[jit] fix side-effects and aliasing for custom ops

Previously we didn't track aliasing, mutation, or side effects for
custom ops. This PR adds in guards with the most conservative
assumptions possible: the op will 1) have side effects, 2) write to
everything, and 3) produce a wildcard. We can reign in those
restrictions later, but want to get something correct for now.

It's a little weird because we have no way of knowing whether a given
node represents a custom op--I added it to the schema but I'm open to
suggestions for a better place.

Follow-ups to this:
- Users should be able to specify aliasing annotations/side effects to
get better perf somehow.
- Relax assumptions a bit.

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

eellison commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -146,12 +155,17 @@ struct FunctionSchema {
-          const auto& aliasInfo = arg.alias_info();
-          return aliasInfo && aliasInfo.value().isWrite();
-        });
+    return is_custom_op_ ||
```
Nit: is the intention that the custom operator returns true even if there are no arguments ? if not, than the is_custom_op_ isn't needed. Otherwise maybe add a comment?

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

eellison commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -682,6 +686,19 @@ void AliasDb::analyzeSetAttr(Node* node) {
+// Custom ops may write to any input and produce wildcards
+void AliasDb::analyzeCustomOp(Node* node) {
+  for (const auto input : node->inputs()) {
+    registerWrite(input, node);
```
Could you add a check in `registerWrite` that the value needs annotation ? Otherwise we'll be disabling constant propagation on primitive values used in custom ops.

eellison commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -126,6 +128,12 @@ struct FunctionSchema {
+  // a schema without alias annotations, so we need to:
+  //   1) know that the schema information is incomplete
+  //   2) treat the op conservatively w.r.t. aliasing and mutation
+  const bool has_alias_annotation_;
```
nit: has_alias_annotation is misleading because unschematized ops don't have alias annotation either

suo replied:
good point. I'll figure out a better name

eellison commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -146,12 +154,17 @@ struct FunctionSchema {
-          const auto& aliasInfo = arg.alias_info();
-          return aliasInfo && aliasInfo.value().isWrite();
-        });
+    return has_alias_annotation_ ||
```
Nit: Is the intention here that unschematized user ops will return true even if they don't have any aliasing inputs?  If yes, add comment ? if no, don't need the `has_alias_annotation_` case

eellison commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -146,12 +154,17 @@ struct FunctionSchema {
-          return aliasInfo && aliasInfo.value().isWrite();
-        });
+    return has_alias_annotation_ ||
+        std::any_of(
```
Above has_alias_annotation says: `False if this schema was inferred for a custom operator`
Is this supposed to be `return !has_alias_annotation  || ...`

suo replied:
sorry I was updating this a bunch. It should be `!has_alias_annotation`.

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

eellison commented on test/cpp/jit/test_custom_operators.h:
```diff
@@ -207,6 +210,53 @@ void testCustomOperators() {
+    const auto text = R"IR(
+graph(%x: Tensor, %y: Tensor):
+  # CHECK: foo::aliasing
+  %ret : Tensor = foo::aliasing(%x, %y)
```
Can you add a test for when there are no inputs or outputs, to test the hasSideEffects() case in addition to the case where we assume inputs are written to?

suo replied:
Apparently the schema inference breaks if you try to return void?

eellison replied:
Could also add a function with no inputs and an unused int output to test the same condition

eellison reviewed (commented):
Have a few comments but is just about ready I think

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

suo commented:
Okay, comments are addressed. This is only a short term fix to get things to be correct—I think in the long run, we should just use the options that #18589 introduces.

eellison commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -146,12 +154,17 @@ struct FunctionSchema {
-          const auto& aliasInfo = arg.alias_info();
-          return aliasInfo && aliasInfo.value().isWrite();
-        });
+    return is_inferred_from_custom_op_ ||
```
So to be clear we are saying that it is mutable even if it doesn't have any inputs, add comment maybe ?

eellison reviewed (approved):
Looks good!

apaszke commented:
Can we just assume that `aten::`/`prim::`/... namespaces are not custom, everything else is?

suo commented:
> Can we just assume that `aten::`/`prim::`/... namespaces are not custom, everything else is?

That's an interesting idea…I don't think namespaces are restricted right now for custom ops, but it's easy to add. And we already have this distinction when emitting builtins.

apaszke commented:
Yeah, we could expose a special subclass of `RegisterOperator` to the users, which disallows registering ops under those namespaces.

eellison commented:
> Can we just assume that `aten::`/`prim::`/... namespaces are not custom, everything else is?

This would require changing some of our tests which use the `test` namespace (probably among others) and in general I think would be hard to maintain

suo commented:
@eellison I think the idea is to designate "reserved" namespaces that users can't define custom ops in. I think this is good idea anyway (right now I think you could create an `aten` op that would participate in builtin schema resolution). Conveniently, we can also use the fact that a op has a non-reserved namespace to denote whether it's custom or not

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

eellison reviewed (dismissed):
Different implementation, have to re-review

suo commented on test/cpp/jit/test_custom_operators.h:
```diff
@@ -206,6 +209,77 @@ void testCustomOperators() {
+  }
+  {
+    RegisterOperators reg(
+        {createOperator("foo::no_input", []() -> double { return 4.0; })});
```
this test causes windows build to fail

eellison commented:
I think the most recent commit will make it harder to move forward to a future api where users declare ops as functional, because we will have ops that are in user space which we do want to assume mutate the inputs.

Operators in c10/caffe2/TVM will also no longer be able to create an operator with specified aliasing, because the namespace will override any information that has been given.

suo commented:
Right now there is no way (short of writing alias annotations manually) for the user to specify the purity of their custom op, so we need to insert guards to make sure things are correct. I expect that when we give users the ability to specify purity, the guards will have to change.

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

zdevito commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -146,12 +146,17 @@ struct FunctionSchema {
+              return aliasInfo && aliasInfo.value().isWrite();
+            });
   }
+  bool is_custom_op() const {
```
For this PR, I think it would be safer to white-list `prim` and `aten` into the stronger alias behavior. I do not like the idea of making a publically-exposed distinction between custom ops and our ops, even for just a few commits.

zdevito commented on aten/src/ATen/core/interned_strings.h:
```diff
@@ -246,6 +246,10 @@ struct CAFFE2_API Symbol {
+  // Does this symbol live in one of the above built-in namespaces?
+  // If not, that means that this symbol came from the user, through the
+  // defintiion of a custom operator.
+  bool is_builtin_ns() const;
```
Please do not commit this. I do not want to have people be able to easily special case stuff for built in ops. This will create huge long-term tech debt even if commit for a little while.

zdevito commented on torch/csrc/jit/custom_operator.h:
```diff
@@ -180,6 +180,12 @@ Operator createOperator(
       std::tuple_size<ArgumentTuple>::value;
 
   auto schema = torch::jit::detail::inferAndCheckSchema<Traits>(schemaOrName);
+  Symbol name = Symbol::fromQualString(schema.name());
```
I don't like this. I want to move our operator registration over to this API and this guard will prevent it. This will need to be removed in the future commit.

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -846,9 +846,14 @@ bool Node::hasSideEffects() const {
+    case prim::TimePoint:
       return true;
   }
+  // Custom ops may have arbitrary side effects
```
```
if (kind.is_aten() || kind.is_print())
 return false
```

zdevito reviewed (changes requested):
Code seems bug free, but see inline comments. I strongly suggest we switch this to whitelist aten and prim rather than special-case all primitive ops.

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

[jit] fix side-effects and aliasing for custom ops

Previously we didn't track aliasing, mutation, or side effects for
custom ops. This PR adds in guards with the most conservative
assumptions possible: the op will 1) have side effects, 2) write to
everything, and 3) produce a wildcard. We can reign in those
restrictions later, but want to get something correct for now.

It's a little weird because we have no way of knowing whether a given
node represents a custom op--I added it to the schema but I'm open to
suggestions for a better place.

Follow-ups to this:
- Users should be able to specify aliasing annotations/side effects to
get better perf somehow.
- Relax assumptions a bit.

gh-metadata: pytorch pytorch 18711 gh/suo/15/head

suo committed:
Update on "[jit] fix side-effects and aliasing for custom ops"

[jit] fix side-effects and aliasing for custom ops

Previously we didn't track aliasing, mutation, or side effects for
custom ops. This PR adds in guards with the most conservative
assumptions possible: the op will 1) have side effects, 2) write to
everything, and 3) produce a wildcard. We can reign in those
restrictions later, but want to get something correct for now.

It's a little weird because we have no way of knowing whether a given
node represents a custom op--I added it to the schema but I'm open to
suggestions for a better place.

Follow-ups to this:
- Users should be able to specify aliasing annotations/side effects to
get better perf somehow.
- Relax assumptions a bit.

gh-metadata: pytorch pytorch 18711 gh/suo/15/head
